feat(outputs.sql): Add option to automate table schema updates

hautecodure opened the pull request:
Automate table updates, when the table does not contain a field

## Summary

Currently, updates to the Table are left to the administrator and whenever a field is not present (in the table), the entire Telegraf flush is discarded, losing data in the process.

The new approach exposes two templates, one to update a table and another on to query a table for it's columns.

To avoid unnecessary (and expensive) requests, we cache the table columns and update it, whenever there is a cache miss from a field.

## Checklist

- [X] No AI generated code was used in this PR

## Related issues

N/A

hautecodure commented:
!signed-cla

hautecodure commented:
Pushed a new version to address the linter complaints and took the opportunity to:

- Get the defer outside of the for loop
- Move the table creation in a separate function to make the Write loop to improve readability

Note: left the `table_exists_template` option, even though nothing uses it, as removing it might break existing configs

srebhan commented on plugins/outputs/sql/sample.conf:
```diff
@@ -25,6 +25,17 @@
   ##  {TABLE} - tablename as a quoted identifier
   # table_exists_template = "SELECT 1 FROM {TABLE} LIMIT 1"
 
+  ## Table update template
+  ## Available template variables:
+  ##  {TABLE} - table name as a quoted identifier
+  ##  {COLUMNS} - column definitions (list of quoted identifiers and types)
+  # column_template = "ALTER TABLE {TABLE} ADD COLUMN IF NOT EXISTS {COLUMN}"
+
+  ## Table columns query template
+  ## Available template variables:
+  ##  {TABLE} - table name as a quoted identifier
+  # column_query_template = "SELECT column_name FROM INFORMATION_SCHEMA.COLUMNS WHERE TABLE_NAME={TABLE}"
```
How about
```suggestion
  ## Table update template, available template variables:
  ##  {TABLE} - table name as a quoted identifier
  ##  {COLUMNS} - column definitions (list of quoted identifiers and types)
  # table_update_template = "ALTER TABLE {TABLE} ADD COLUMN IF NOT EXISTS {COLUMN}"

  ## Table columns query template, available template variables:
  ##  {TABLE} - table name as a quoted identifier
  # table_column_query_template = "SELECT column_name FROM INFORMATION_SCHEMA.COLUMNS WHERE TABLE_NAME={TABLE}"
```

srebhan commented on plugins/outputs/sql/README.md:
````diff
@@ -183,6 +193,16 @@ It is not supported on windows/386, mips, and mips64 platforms.
 The DSN is a filename or url with scheme "file:". See the [driver
 docs](https://modernc.org/sqlite) for details.
 
+As SQLite does not support `INFORMATION_SCHEMA`, you'll have to update
+`column_template` to something like:
+
+```toml
+# Save metrics to an SQL Database
+[[outputs.sql]]
+  ## ...
+  column_query_template = "SELECT name AS column_name FROM pragma_table_info({TABLE})"
+```
+
````
This concerns me to be honest. My fear that there are more SQL databases out there not supporting the column query or altering the table. Furthermore, there might be databases requiring special permissions to alter/query the table columns. In all of those cases we do break existing setups...

So how about making this opt-in? I.e. you leave the template empty by default and add your current default as an example to the `sample.conf` and/or README. This way, you can skip the table updates if the template(s) are empty. What do you think?

hautecodure replied:
As of now, based on the drivers currently in use by the plugin, only SQLite does not support this method of querying the database for its columns.

- List of databases, that support [Information schema](https://en.wikipedia.org/wiki/Information_schema)
- List of SQL drivers for [database/sql](https://go.dev/wiki/SQLDrivers)

Will change to opt-in, though as it automates the setup, i think at some point we should change it opt-out, and get people to upgrade.

The reason why i didn't consider opt-in/out was the assumption that it's worth it and prevents errors, due to new tags / metrics present in the buffer.

srebhan replied:
Ok so I see two ways:
1. We have an _opt-in_ and this will likely stay for a long-time but we are sure the plugin does not break existing setups.
2. We remove the setting altogether and implement this function as an always-on feature.

For the latter to happen we need to
- Make sure we don't dramatically increase the number of queries issued against the database.
- Implement integration tests for all/most-important supported drivers to verify we don't break things.

Honestly speaking, I expect most people to setup the schema of their database and don't want it to be updated by some random agent...

hautecodure replied:
> We have an opt-in and this will likely stay for a long-time but we are sure the plugin does not break existing setups.

Went with opt-in, as impacting existing users is not the way to go.

> Make sure we don't dramatically increase the number of queries issued against the database.

The queries would be at least >2x for each metric (plugin). One for the table, one for the columns. However, these are one-time queries (as they are cached) and spread-out over time.

But >2x is not... much. If a user has 50 enabled plugins, they are looking at making 100+ SELECT queries, when the next flush comes around, which is... not much.

In fact, most of these queries will be to INFORMATION_SCHEMA.{columns,tables} tables, which are fine-tuned and do not contain that many rows to affect the throughput or latency of the database.

Worst case, the database is slow and struggles to process these queries: once the results are cached, there's is not going to be any impact on the performance.

> Honestly speaking, I expect most people to setup the schema of their database and don't want it to be updated by some random agent...

You may be right, i just found it tedious, going through each plugin and figuring out which fields/tags i should create and what should be their type. i just wanted to automate it and this may be skewing my perception, when it comes to usefulness.

srebhan commented on plugins/outputs/sql/README.md:
```diff
@@ -106,6 +105,17 @@ See the [CONFIGURATION.md][CONFIGURATION.md] for more details.
+  ##  {COLUMNS} - column definitions (list of quoted identifiers and types)
+  # column_template = "ALTER TABLE {TABLE} ADD COLUMN IF NOT EXISTS {COLUMN}"
+
+  ## Table columns query template
+  ## Available template variables:
+  ##  {TABLE} - table name as a quoted identifier
+  # column_query_template = "SELECT column_name FROM INFORMATION_SCHEMA.COLUMNS WHERE TABLE_NAME={TABLE}"
```
Is there any reason the user would change that? I mean if we know those templates, why should the user configure them?

hautecodure replied:
Depending on the database, the query might (does) look different. For example we had to modify sqlite_test.go (part of this PR), as SQLite does not support INFORMATION_SCHEMA.

srebhan replied:
Right, but we know the query for each driver type don't we? So we could setup the correct query internally, taking away the burden from the user...

hautecodure replied:
Ok, will remove it and leave a comment for devs who want to add a driver to the plugin.

srebhan commented on plugins/outputs/sql/sql.go:
```diff
@@ -200,28 +213,130 @@ func (p *SQL) generateInsert(tablename string, columns []string) string {
+	stmt := p.ColumnQueryTemplate
+	stmt = strings.ReplaceAll(stmt, "{TABLE}", quoteStr(tableName))
+
+	columns, err := p.db.Query(stmt)
+
+	if err != nil {
+		return err
+	}
+
+	defer columns.Close()
```
Please do not randomly insert empty lines! They should be used sparsely as elements to separate logical operation blocks.
```suggestion
	columns, err := p.db.Query(stmt)
	if err != nil {
		return err
	}
	defer columns.Close()
```
Same for other instances below.

srebhan commented on plugins/outputs/sql/sql.go:
```diff
@@ -200,28 +213,130 @@ func (p *SQL) generateInsert(tablename string, columns []string) string {
+}
+
+func (p *SQL) createTableIfDoesNotExist(metric telegraf.Metric) error {
+	createStmt := p.generateCreateTable(metric)
+	_, err := p.db.Exec(createStmt)
+	if err != nil {
+		return err
+	}
+
+	err = p.fetchColumns(metric.Name())
+	if err != nil {
+		return err
+	}
+
+	return nil
```
Using implicit declaration and short code is nicer to read as you don't need to worry about the scope. Furthermore, it would be good to add some error prefixes to easier spot the issue when debugging user errors later.
```suggestion
	stmt := p.generateCreateTable(metric)
	if _, err := p.db.Exec(stmt); err != nil {
		return fmt.Errorf("creating table failed: %w", err)
	}

	if err := p.fetchColumns(metric.Name()); err != nil {
		return fmt.Errorf("fetching columns failed: %w", err)
	}

	return nil
```

srebhan commented on plugins/outputs/sql/sql.go:
```diff
@@ -200,28 +213,130 @@ func (p *SQL) generateInsert(tablename string, columns []string) string {
+			return err
+		}
+
+		err = p.fetchColumns(tablename)
+
+		if err != nil {
+			return err
+		}
```
Why do you fetch the columns again?

hautecodure replied:
fetchColumns populates the cache, we could update it directly, but that way we would duplicate code.

srebhan replied:
Yeah what I meant is that in the previous version you also fetched the column when updating the table...

srebhan commented on plugins/outputs/sql/sql.go:
```diff
@@ -200,28 +213,130 @@ func (p *SQL) generateInsert(tablename string, columns []string) string {
-		if !p.tables[tablename] && !p.tableExists(tablename) {
-			createStmt := p.generateCreateTable(metric)
-			_, err := p.db.Exec(createStmt)
+		if _, isTableCached := p.tables[tablename]; !isTableCached {
+			err := p.createTableIfDoesNotExist(metric)
+
```
```suggestion
		if _, found := p.tables[tablename]; !found {
			if err := p.createTableIfDoesNotExist(metric); err != nil {
```

srebhan commented on plugins/outputs/sql/sql.go:
```diff
@@ -234,43 +349,40 @@ func (p *SQL) Write(metrics []telegraf.Metric) error {
-			if err != nil {
-				return fmt.Errorf("execution failed: %w", err)
-			}
+		// sort the fields to improve the "Prepared" statement throughput
+		sortedColumns, sortedParams := p.sortFields(columns, values)
+
+		sql := p.generateInsert(tablename, sortedColumns)
+		batchMetrics[sql] = append(batchMetrics[sql], sortedParams)
+	}
+
+	for sql, sqlParams := range batchMetrics {
+		err := p.sendBatchTx(sql, sqlParams)
+
+		if err != nil {
+			return err
```
I think this should be stripped out to another PR as it is in no way related to adding the columns...

hautecodure replied:
Sure

srebhan reviewed (changes requested):
@hautecodure thank you very much for your contribution! Much appreciated! I do have some comments in the code, the biggest work items are
1. Strip out the column sorting to another PR to only have one thing in each PR!
2. I would rather like to have table-altering as an opt-in to not accidentally break existing configurations.

Looking forward to the next round!

srebhan commented on plugins/outputs/sql/sql.go:
```diff
@@ -211,14 +277,11 @@ func (p *SQL) Write(metrics []telegraf.Metric) error {
-			createStmt := p.generateCreateTable(metric)
-			_, err := p.db.Exec(createStmt)
-			if err != nil {
+		if _, found := p.tables[tablename]; !found {
+			if err := p.createTableIfDoesNotExist(metric); err != nil {
```
Aren't you generating the tables unconditionally now?

hautecodure replied:
Right, i was using "IF NOT EXIST" clause, which makes the subsequent query redundant. The only issue is that Microsoft SQL Server does not support it (every other driver does).

I thought about hard-coding it (by adding a workaround, when microsoft sql is in use), but the plugin starts to look like kinda bad, with driver-conditions in every other function.

I think using the tableExist function (for now) seems to be the most compatible approach.

hautecodure commented:
Pushed a new update:

- Added tests
- Reworked the table query, as it did not work under Clickhouse.
- The column query has been removed
- Changed a few if/else expression to match the style convention

While a bit unpleasant with the per-driver overrides, there was not much of a choice, unless we want to pester users with writing queries for each and every edge-case.

srebhan reviewed (commented):
Thanks @hautecodure for the update and your patience! I do have some styling comments in the code. Additionally, I have the following items:
1. Please do **not** remove existing options (`table_exists_template`) or render them non-functional! This might break some users! I agree that most likely this option is useless and should be removed, but to do so, you need to deprecate it first and keep it for a certain time.
2. The default value of `table_column_template` is empty and therefore, the line in the `sample.conf` file must reflect this! You can add a line
```toml
  ## Use the following setting for automatically adding columns:
  ## table_update_template = "ALTER TABLE {TABLE} ADD COLUMN {COLUMN}"
```
3. Using `tables map[string]map[string]bool` might simplify the code and improve lookups. What do you think?

hautecodure commented on plugins/outputs/sql/sql.go:
```diff
@@ -197,11 +209,98 @@ func (p *SQL) generateInsert(tablename string, columns []string) string {
 }
 
+func (p *SQL) updateTableCache(tablename string) error {
+	var stmt string
+
+	switch p.Driver {
+	case "sqlite":
+		stmt = "SELECT name AS column_name FROM pragma_table_info({TABLE})"
+	default:
+		stmt = "SELECT column_name FROM INFORMATION_SCHEMA.COLUMNS WHERE TABLE_NAME={TABLE}"
+	}
+
+	stmt = strings.ReplaceAll(stmt, "{TABLE}", quoteStr(tablename))
```
I noticed in an earlier Clickhouse PR, a recommendation to use switch as new drivers may be added in the future.

Should i move to an if statement in tableExists as well?

hautecodure commented on plugins/outputs/sql/sql.go:
```diff
@@ -197,11 +209,98 @@ func (p *SQL) generateInsert(tablename string, columns []string) string {
+		if _, err := p.db.Exec(stmt); err != nil {
+			return fmt.Errorf("creating table failed: %w", err)
+		}
+		if err := p.updateTableCache(tablename); err != nil {
+			return fmt.Errorf("updating table cache failed: %w", err)
+		}
```
Right... guess i'll have to restore tableExist and it's options as users may or may not have permissions to query these tables.

hautecodure commented on plugins/outputs/sql/sql.go:
```diff
@@ -51,7 +53,7 @@ type SQL struct {
 
 	db     *gosql.DB
-	tables map[string]bool
+	tables map[string][]string
```
Thought that slices is a bit cleaner, but you're right. Slices.Contains in the hot-path is not a great idea

hautecodure commented:
Hi, thanks for the feedback, pushed a new iteration:

- Reverted `tableExists` back to its current form - not in the scope for this PR, might open a new one, to discuss / update (if needed).
- Update the SQL struct names, to closely match the config options
- Switch table cache to `map[string]map[string]bool`

srebhan commented on plugins/outputs/sql/sql.go:
```diff
@@ -201,28 +211,95 @@ func (p *SQL) generateInsert(tablename string, columns []string) string {
+		return nil
+	}
+
+	stmt := strings.ReplaceAll(p.TableListColumnsTemplate, "{TABLE}", quoteStr(tablename))
+
+	columns, err := p.db.Query(stmt)
+	if err != nil {
+		return fmt.Errorf("fetching columns for table(%s) failed: %w", tablename, err)
+	}
+	defer columns.Close()
+
+	for columns.Next() {
+		var columnName string
+		if err := columns.Scan(&columnName); err != nil {
+			return err
+		}
+
+		if !p.tables[tablename][columnName] {
+			p.tables[tablename][columnName] = true
+		}
+	}
+
+	return nil
+}
+
 func (p *SQL) Write(metrics []telegraf.Metric) error {
 	var err error
 
 	for _, metric := range metrics {
 		tablename := metric.Name()
 
 		// create table if needed
-		if !p.tables[tablename] && !p.tableExists(tablename) {
-			createStmt := p.generateCreateTable(metric)
-			_, err := p.db.Exec(createStmt)
-			if err != nil {
-				return err
-			}
+		if err := p.createTable(metric); err != nil {
+			return err
```
To be honest, I would like to keep the _condition_ here as otherwise this reads like "always create the table"!

hautecodure replied:
I was trying to keep the Write loop easier to read and handle the minutia in the respective function.

srebhan replied:
Yes, please revert back. My reasoning is that we can directly see in the write loop that tables are _only_ created if they do not exist. Jumping to the `createTable` function only to see it directly returns is disturbing when debugging code...

srebhan commented on plugins/outputs/sql/sql.go:
```diff
@@ -201,28 +211,95 @@ func (p *SQL) generateInsert(tablename string, columns []string) string {
 }
 
+func (p *SQL) updateTableCache(tablename string) error {
+	// modifying the table schema is opt-in
+	if p.TableUpdateTemplate == "" {
+		return nil
+	}
```
Can we please move the condition to the caller place to make clear that we only update if opted-in!? It's more a code reading thing than a functional change...

srebhan commented on plugins/outputs/sql/sql.go:
```diff
@@ -201,28 +211,95 @@ func (p *SQL) generateInsert(tablename string, columns []string) string {
+}
+
+func (p *SQL) createColumn(tablename, column, columnType string) error {
+	// modifying the table schema is opt-in
+	if p.TableUpdateTemplate == "" {
+		return nil
+	}
```
Can we please move the condition to the caller place to make clear that we only update if opted-in!? It's more a code reading thing than a functional change...

srebhan commented on plugins/outputs/sql/sql_test.go:
```diff
@@ -288,7 +337,6 @@ func TestPostgresIntegration(t *testing.T) {
 	require.NoError(t, p.Connect())
 	defer p.Close()
 	require.NoError(t, p.Write(testMetrics))
-	require.NoError(t, p.Close())
```
Why do you remove this? This line makes sure all metrics are written...

hautecodure replied:
It closes the connection and i want to re-use the existing connection, instead of setting up a new one.

The tests seem to pass fine. Are you sure Close is required to flush the metrics? Is this a pgx quirk?

srebhan replied:
Yes this is on purpose! If `Close` finishes we know all commands are flushed to the server and as I said below, please create an own test-case for your stuff!

srebhan commented on plugins/outputs/sql/sql_test.go:
```diff
@@ -316,6 +364,40 @@ func TestPostgresIntegration(t *testing.T) {
 
 		return bytes.Contains(b, expected)
 	}, 5*time.Second, 500*time.Millisecond)
+
+	p.TableUpdateTemplate = "ALTER TABLE {TABLE} ADD COLUMN {COLUMN}"
+
+	require.NoError(t, p.Write(postCreateMetrics))
+
+	fields := []string{
+		"tag_add_after_create text",
+		"bool_add_after_create boolean",
+	}
+	for _, column := range fields {
+		require.Eventually(t, func() bool {
+			rc, out, err := container.Exec([]string{
+				"bash",
+				"-c",
+				"pg_dump" +
+					" --username=" + username +
+					" --no-comments" +
+					" " + dbname +
+					// pg_dump's output has comments that include build info
+					// of postgres and pg_dump. The build info changes with
+					// each release. To prevent these changes from causing the
+					// test to fail, we strip out comments. Also strip out
+					// blank lines.
+					"|grep -E -v '(^--|^$|^SET )'",
+			})
+			require.NoError(t, err)
+			require.Equal(t, 0, rc)
+
+			b, err := io.ReadAll(out)
+			require.NoError(t, err)
+
+			return bytes.Contains(b, []byte(column))
+		}, 5*time.Second, 500*time.Millisecond, column)
+	}
```
Please create a new testcase for this or directly set the update template! Furthermore, a comment would be nice saying something like
```golang
	// Write a metric that targets the same table but has additional fields to test the automatic
	// column update functionality.
	require.NoError(t, p.Write(postCreateMetrics))

	fields := []string{
		"tag_add_after_create text",
		"bool_add_after_create boolean",
	}
	for _, column := range fields {
```
Same for the other spots below.

srebhan replied:
Simply duplicate the existing test and add your code in a new `TestMysqlUpdateSchemeIntegration` function. There is no need to squash all of them into a single test. Having separate functions also allows to specifically run your test!

srebhan reviewed (commented):
Thanks @hautecodure. Some more cosmetic things like moving the conditions to the caller side so that the reader directly sees that the create/update functions are only called on purpose or opt-in.

hautecodure commented:
Thanks for sticking around for another iteration on this PR, rebased and updated to address the issues raised during the last review.

srebhan commented on plugins/outputs/sql/sql.go:
```diff
@@ -48,21 +48,23 @@ type ConvertStruct struct {
+	TableTemplate            string          `toml:"table_template"`
+	TableExistsTemplate      string          `toml:"table_exists_template"`
+	TableUpdateTemplate      string          `toml:"table_update_template"`
+	TableListColumnsTemplate string          `toml:"-"`
```
Do not export this (i.e. start with lowercase letter) and move it down to the internal variables.

srebhan commented on plugins/outputs/sql/sample.conf:
```diff
@@ -29,6 +29,15 @@
 
+  ## Table update template, available template variables:
+  ##  {TABLE} - table name as a quoted identifier
+  ##  {COLUMNS} - column definitions (list of quoted identifiers and types)
```
Is this really `{COLUMNS}` or `{COLUMN}`?

Hipska replied:
It's a typo.
```suggestion
  ##  {COLUMN} - column definition (quoted identifier and type)
```

srebhan commented on plugins/outputs/sql/sql.go:
```diff
@@ -278,6 +351,15 @@ func (p *SQL) Write(metrics []telegraf.Metric) error {
 
+		// modifying the table schema is opt-in
+		if p.TableUpdateTemplate != "" {
+			for i := range len(columns) {
+				if err := p.createColumn(tablename, columns[i], p.deriveDatatype(values[i])); err != nil {
+					return err
+				}
+			}
```
Would it make sense to create all columns in one statement to reduce the number of queries...

hautecodure replied:
I had an earlier version doing this, but there was a major issue:

Unlike CREATE TABLE, you specify the options as an ADD COLUMN directive and repeat for every column

This complicates the string template, as parts of the query have to be repeated. And the user is required to enter a sort-of SQL query, which might end up more complex and confusing.

As far the number of queries goes... it's a hard one to gauge.

- Most plugins do not create new fields/tags
- ...even if they do, it's just a few new fields per flush

To take advantage and ensure there is >1 new field in a single flush:

- the user must use the merge aggregator
- as well as buffer more metrics, so there are more fields to process

Which is why i wasn't sure if it's worth it and ended up removing it.

srebhan replied:
Ok, let's leave this optimization for later then!

srebhan reviewed (commented):
Sorry man but some more questions and requests...

Hipska commented on plugins/outputs/sql/README.md:
```diff
@@ -155,6 +164,29 @@ See the [CONFIGURATION.md][CONFIGURATION.md] for more details.
+If you wish the plugin to sync the column-schema for every metric,
+uncomment the `table_update_template` setting in your config file.
+
+> [!NOTE] The default setting is a generic query that your database may
+> or may not support. Consult your database's documentation for proper
+> syntax and table options.
```
I don't see a default defined in the code, so this sentence isn't correct.

hautecodure replied:
Thanks for heads-up, indeed it was a poorly worded and confusing note. Pushed a new version, let me know if it still sounds a bit off.

Hipska commented on plugins/outputs/sql/README.md:
```diff
@@ -155,6 +164,29 @@ See the [CONFIGURATION.md][CONFIGURATION.md] for more details.
+subsequent metrics with additional fields or tags will result in errors.
+
+If you wish the plugin to sync the column-schema for every metric,
+uncomment the `table_update_template` setting in your config file.
```
```suggestion
specify the `table_update_template` setting in your config file.
```

srebhan commented:
@hautecodure could you please rebase on latest master to get CI fixed!?

Hipska commented on plugins/outputs/sql/README.md:
````diff
@@ -155,6 +164,29 @@ See the [CONFIGURATION.md][CONFIGURATION.md] for more details.
+```toml
+# Save metrics to an SQL Database
+[[outputs.sql]]
+  ## Table update template, available template variables:
+  ##  {TABLE} - table name as a quoted identifier
+  ##  {COLUMNS} - column definitions (list of quoted identifiers and types)
+  ## NOTE: Ensure the user (you're using to write to the database) has necessary permissions
````
Maybe simplify it even more, as it already is inconsistent with the sample config:
```suggestion
  ## Table update template
```

hautecodure replied:
Thanks, somehow missed this one

srebhan reviewed (approved):
Awesome! Thanks @hautecodure!

skartikey commented on plugins/outputs/sql/sql.go:
```diff
@@ -213,6 +220,14 @@ func (p *SQL) generateCreateTable(metric telegraf.Metric) string {
 	return query
 }
 
+func (p *SQL) generateAddColumn(tablename, column, columnType string) string {
+	query := p.TableUpdateTemplate
+	query = strings.ReplaceAll(query, "{TABLE}", quoteIdent(tablename))
+	query = strings.ReplaceAll(query, "{COLUMN}", quoteIdent(column)+" "+columnType)
+
+	return query
+}
```
Instead of performing multiple `strings.ReplaceAll` calls, `fmt.Sprintf` might be more readable and efficient.
```suggestion
func (p *SQL) generateAddColumn(tablename, column, columnType string) string {
	table := quoteIdent(tablename)
	col := quoteIdent(column)

	return fmt.Sprintf(p.TableUpdateTemplate, table, col+" "+columnType)
}

```

skartikey replied:
Please leave it as it is, then. Thanks!

skartikey commented on plugins/outputs/sql/sql.go:
```diff
@@ -237,28 +252,86 @@ func (p *SQL) generateInsert(tablename string, columns []string) string {
+		return fmt.Errorf("fetching columns for table(%s) failed: %w", tablename, err)
+	}
+	defer columns.Close()
+
```
wondering if table doesn't exist, does it make sense to initialise the map?
```
// Initialize the map if it doesn't exist
    if p.tables[tablename] == nil {
        p.tables[tablename] = make(map[string]bool)
    }
```

hautecodure replied:
Right now it shouldn't be an issue, but a code refactor might make it easy to assume that it's a self-contained function. I'll update the function as suggested.

hautecodure commented on plugins/outputs/sql/sql.go:
```diff
@@ -237,28 +252,86 @@ func (p *SQL) generateInsert(tablename string, columns []string) string {
+	return nil
+}
+
+func (p *SQL) createColumn(tablename, column, columnType string) error {
+	// update table cache with the latest table schema
+	// some servers may lack support for "IF NOT EXIST"
+	// clause in which case, we don't want to burden users
+	// with writing complex "IF NOT EXIST" workarounds
+	if !p.tables[tablename][column] {
+		if err := p.updateTableCache(tablename); err != nil {
+			return fmt.Errorf("updating table cache failed: %w", err)
+		}
+	}
+	if !p.tables[tablename][column] {
+		createColumn := p.generateAddColumn(tablename, column, columnType)
+		if _, err := p.db.Exec(createColumn); err != nil {
+			return fmt.Errorf("creating column failed: %w", err)
+		}
+		if err := p.updateTableCache(tablename); err != nil {
+			return fmt.Errorf("updating table cache failed: %w", err)
+		}
+	}
+	return nil
+}
```
> Additionally, p.tables[tablename][column] assumes tablename exists in p.tables, which may cause a panic if it does not.

It shouldn't be an issue as the table cache is populated in the Write loop, before invoking createColumn.

> A more efficient approach would be to update the table cache first and then check the column’s existence only once.

I don't mind adding more guard rails, I'll update the function as suggested
